### DISCUSSION TITLE
using-scoped-services-inside-a-quartz-net-hosted-service-with-asp-net-core/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#97] [Announcements]
# using-scoped-services-inside-a-quartz-net-hosted-service-with-asp-net-core/

In this post I describe how to run Quartz.NET jobs using an ASP.NET Core hosted service, and how to handle scoped services inside singleton jobs.

https://andrewlock.net/using-scoped-services-inside-a-quartz-net-hosted-service-with-asp-net-core/

**Comment by xantari** [comment]
There's a problem in the last code snippet I can't figure out. If you do not at [DisallowConcurrentExecution] on the IJob implementation you will get threading issues with the DbContext which yields the following error:

A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext

A quick example of this is to schedule the same job with two different triggers that execute at the same time (I did every 1 minute Cron as a test: 0 * * ? * * *

Have not been able to figure out a way around this at this time, which is unfortunate as it slows down my background task considerably.

**Reply by andrewlock** [reply]
Oops, sorry for the delay! You're right, there is a problem! It should look like this:

```diff
-            var dbContext = _serviceProvider.GetRequiredService<AppDbContext>();
+            var dbContext = scope.ServiceProvider.GetRequiredService<AppDbContext>();
-            var messageBus = _serviceProvider.GetRequiredService<IBus>();
+            var messageBus = scope.ServiceProvider.GetRequiredService<IBus>();
```

**Reply by xantari** [reply]
Sweet, yeah I ultimately figured that part out. Thanks for pointing this stuff out in your blog!

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 30 2019, 06:46, in reply to [devbob](https://disqus.com/by/devbob/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/97#discussioncomment-6056602)</em>

---

<p>Hi, yes you're absolutely right with your changes, it should be a Singleton, thanks! :)</p>

[a comment by an automated account is not shown]

**Comment by dariogriffo** [comment]
As of Quartz.net 3.3.2 all jobs are scoped by default. Doc [here](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/microsoft-di-integration.html#di-aware-job-factories)
